module/extension/Director/src/bits/challenge_priority.cpp

```diff
@@ -0,0 +1,104 @@
+
+            uint64_t id;
+            int priority;
+            bool optional;
+        };
+    }  // namespace
+
+    bool Director::challenge_priority(const std::shared_ptr<const DirectorTask>& incumbent,
+                                      const std::shared_ptr<const DirectorTask>& challenger) {
+
+        // Compare to yourself and you get that you shouldn't change tasks
+        if (incumbent->requester_id == challenger->requester_id) {
+            return false;
+        }
+
+        // Function to get the priorities of the ancestors of this task
+        auto get_ancestor_priorities = [this](const std::shared_ptr<const DirectorTask>& task) {
+            // We are our first anscestor
+            std::vector<TaskPriority> ancestors;
+            ancestors.emplace_back(task->requester_id, task->priority, task->optional);
+
+            // Loop up through the providers until we reach a point where a task was emitted by a non provider
+            for (auto t = task; providers.count(t->requester_id) != 0;) {
+
+                // Get the provider that emitted this task, and from that the provider group
+                auto provider = providers[t->requester_id];
+                auto& group   = groups[provider->type];
+
+                // If there is no active task something has gone wrong with the algorithm
+                if (group.active_task == nullptr) {
+                    throw std::runtime_error("Task has broken parentage");
+                }
+
+                // Add this new parent we found and set t so we can loop up further
+                t = group.active_task;
+                ancestors.emplace_back(t->requester_id, t->priority, t->optional);
+            }
+            return ancestors;
+        };
+
+        // Convenience references for the two to make the code shorter
+        const auto& a = incumbent;
+        const auto& b = challenger;
+
+        // Get the ancestor priorities of a and b
+        auto a_p = get_ancestor_priorities(a);
+        auto b_p = get_ancestor_priorities(b);
+
+        // Remove all of the common ancestors
+        while (a_p.back().id == b_p.back().id) {
+            a_p.pop_back();
+            b_p.pop_back();
+        }
+
+        // Work out if there are any optionals in either of the tasks parentage
+        const bool a_optional = std::any_of(a_p.begin(), a_p.end(), [](const auto& v) { return v.optional; });
+        const bool b_optional = std::any_of(b_p.begin(), b_p.end(), [](const auto& v) { return v.optional; });
+
+        // If both or neither are optional then we compare at the point where they were siblings.
+        // If both are optional then we would rather that whichever ancestor had higher priority have its optional tasks
+        if (a_optional == b_optional) {
+            return a_p.back().priority < b_p.back().priority;
+        }
+
+        // If the optionals are not equal then return if a_optional is true. a being optional would make it less than b
```

Comment:
This comment doesn't seem to make sense to me - it's not only returning if a_optional is true it's just returning the value of a_optional? I'm guessing "it" in this case you are referring to the priority of task a?

Reply:
I made the comment more involved for you

Reply:
Thankyou for putting up with me being big dum 😁